File: gui/src/lib/studio/fixactions.js
// fixactions.js — the client half of the remediation vocabulary.
//
// The Go half is internal/studio/fixactions.go. Every id it can emit must be
// handled here; every id handled here must exist there. fixactions.test.js
// reads both files and fails the build in either direction, because this seam
// has drifted twice already: Studio.svelte once handled five actions the server
// never sent while six it did send fell through to a no-op, and the security
// panel later grew a second vocabulary of its own.
//
// Everything here is pure. Applying a fix returns a NEW draft and a sentence
// describing what changed; it never touches the DOM, navigates, or reads a
// store. That keeps the interesting half — what the button actually does to
// your workflow — testable without mounting a 10k-line component.

/** Actions that leave Studio for the screen that owns the setting. */
export const NAVIGATE_TARGETS = {
  open_providers: '#providers',
  open_mcp: '#mcp',
  open_delivery: '#channels',
  open_secrets: '#secrets',
}

/** Actions the host handles in place: focus a node, open the bench, etc. */
export const FOCUS_ACTIONS = [
  'choose_model',
  'add_assertions',
  'run_live',
  'open_studio',
  'open_preflight',
  'reveal_node',
  'rename_agent',
]

// Channels that reach people outside the install — the exact set in
// studioSharedExternalChannels (internal/studio/security_preflight.go). The
// first version of this list was hand-typed and got two of them wrong
// ("googlechat" for google_chat, and sms missing entirely), so the button would
// have claimed an agent was internal-only while leaving those two wired up.
// fixactions.test.js now reads the Go map itself rather than a pattern someone
// typed twice.
export const SHARED_CHANNELS = [
  'telegram', 'slack', 'discord',
  'whatsapp', 'whatsapp_web',
  'email', 'teams', 'google_chat',
  'sms', 'webhook',
]

const isShared = (c) => SHARED_CHANNELS.includes(String(c || '').toLowerCase())

/**
 * Fixes Studio applies to the draft itself.
 * Each returns { draft, message } on success, or { message } with no draft when
 * there was nothing to do — the caller reports that rather than silently
 * "succeeding" on a no-op.
 */
export const DRAFT_FIXES = {
  restrict_to_internal_channels(draft) {
    const before = Array.isArray(draft.channels) ? draft.channels : []
    const removed = before.filter(isShared)
    if (!removed.length) {
      return { message: 'This agent is already on internal channels only.' }
    }
    const kept = before.filter((c) => !isShared(c))
    // Leave it reachable. An agent bound to nothing at all is a different kind
    // of broken from an agent bound to too much.
    if (!kept.includes('http')) kept.push('http')

    // The graph's output node routes to channels too. Leaving those behind made
    // the fix a half-measure: the security review still saw the shared channel
    // in the graph and kept warning, while the agent's channel list said it had
    // been dealt with.
    const flow = draft.flow || {}
    const output = flow.output && typeof flow.output === 'object'
      ? { ...flow.output, channels: (flow.output.channels || []).filter((c) => !isShared(c)) }
      : flow.output
    const nodes = Array.isArray(flow.nodes)
      ? flow.nodes.map((n) => (n && Array.isArray(n.channels) && n.channels.some(isShared)
        ? { ...n, channels: n.channels.filter((c) => !isShared(c)) }
        : n))
      : flow.nodes

    // Say the consequence out loud. Removing the only real destination from an
    // agent whose job is delivery turns one warning into two blockers, and the
    // user should hear that from the button rather than from the Save step.
    const stillDelivers = kept.some((c) => !['http', 'webhook', ''].includes(String(c).toLowerCase()))
    const consequence = stillDelivers
      ? ''
      : ' It now has no outbound destination — pick another one, or accept the exposure per binding on the Delivery page instead.'

    return {
      draft: { ...draft, channels: kept, flow: { ...flow, output, nodes } },
      message: `Removed ${removed.join(', ')} — this agent is now on internal HTTP only.${consequence}`,
    }
  },

  // The synthesized persona travels in the finding, so the click is instant and
  // the client never has to know what a good agent prompt looks like.
  write_helper_prompt(draft, params = {}) {
    const id = String(params.agent || '').trim()
    const prompt = String(params.prompt || '').trim()
    if (!id || !prompt) {
      return { message: 'No starter prompt was offered for this helper agent.' }
    }
    const peers = Array.isArray(draft.new_agents) ? draft.new_agents : []
    const i = peers.findIndex((p) => String(p && p.id || '').toLowerCase() === id.toLowerCase())
    if (i < 0) {
      // The peer is referenced by a node but has no profile entry yet, so make
      // one rather than dropping the fix on the floor.
      return {
        draft: { ...draft, new_agents: [...peers, { id, name: id, system_prompt: prompt }] },
        message: `Wrote a starter prompt for ${id}. Read it over — it is a floor, not a ceiling.`,
      }
    }
    const next = peers.slice()
    next[i] = { ...next[i], system_prompt: prompt }
    return {
      draft: { ...draft, new_agents: next },
      message: `Wrote a starter prompt for ${id}. Read it over — it is a floor, not a ceiling.`,
    }
  },

  set_intent_gate_deny(draft) {
    if ((draft.security || {}).intent_gate === 'deny') {
      return { message: 'The intent gate is already set to deny.' }
    }
    return {
      draft: { ...draft, security: { ...(draft.security || {}), intent_gate: 'deny' } },
      message: 'Intent gate set to deny — injection-steered tool calls will be refused.',
    }
  },
}

/** Every id this module knows how to handle. */
export function handledActionIds() {
  return [
    ...Object.keys(NAVIGATE_TARGETS),
    ...FOCUS_ACTIONS,
    ...Object.keys(DRAFT_FIXES),
  ].sort()
}

/** What kind of thing an id is, or '' when it is not in the vocabulary. */
export function actionKind(id) {
  if (!id) return ''
  if (NAVIGATE_TARGETS[id]) return 'navigate'
  if (DRAFT_FIXES[id]) return 'apply'
  if (FOCUS_ACTIONS.includes(id)) return 'focus'
  return ''
}

/**
 * Apply a draft-editing fix. Returns { draft, message } — `draft` is undefined
 * when nothing changed. Throws for an unknown id, because a caller silently
 * dropping one is exactly the failure this module exists to prevent.
 */
export function applyDraftFix(draft, action, params) {
  const fn = DRAFT_FIXES[action]
  if (!fn) throw new Error(`no draft fix for action "${action}"`)
  if (!draft) return { message: 'There is no draft to change yet.' }
  return fn(draft, params || {})
}

/** The button text, when the server did not send one. */
export function fallbackLabel(action) {
  return actionKind(action) ? 'Fix this' : ''
}

File: internal/studio/fixactions.go
package studio

// fixactions.go — the one vocabulary of remediation actions.
//
// Every kind of finding Studio produces — preflight issues, contract checks,
// security findings, readiness items — ends up in front of a user who wants to
// know one thing: what do I click. Prose alone cannot answer that. An id can:
// the client turns it into a button that either lands the user on the screen
// that owns the setting, or edits the draft outright.
//
// The ids and their button text live HERE, in one list, for two reasons.
//
// First, a button's wording is part of the finding, not part of the panel. When
// three panels each kept their own switch, the same action read as "Configure
// provider" in one place and "Fix this" in another.
//
// Second, this is a hand-maintained seam across two languages, and it has
// already drifted twice in this codebase. Studio.svelte's readinessAction once
// handled five actions the server never emitted while six the server did emit
// fell through to a no-op — a "Fix this" button that did nothing at all. The
// security panel then grew its own parallel vocabulary. Both sides were
// internally consistent; nothing checked them against each other.
//
// So: add an action here, and gui/src/lib/studio/fixactions.js must handle it.
// fixactions.test.js reads both files and fails the build in either direction.

import "strings"

// What pressing the button does.
const (
	// FixKindNavigate sends the user to the screen that owns the setting. The
	// change is not Studio's to make — a provider key, a channel binding, an
	// operator's acceptance of risk.
	FixKindNavigate = "navigate"
	// FixKindApply edits the draft in place. Studio owns the value, so the
	// button can just set it.
	FixKindApply = "apply"
	// FixKindFocus keeps the user where they are and moves their attention —
	// reveals a node, opens the bench, switches wizard step.
	FixKindFocus = "focus"
)

// Action ids. These strings cross the wire; treat them as API.
const (
	FixOpenProviders = "open_providers"
	FixOpenMCP       = "open_mcp"
	FixOpenDelivery  = "open_delivery"
	FixOpenSecrets   = "open_secrets"

	FixChooseModel   = "choose_model"
	FixAddAssertions = "add_assertions"
	FixRunLive       = "run_live"
	FixOpenStudio    = "open_studio"
	FixOpenPreflight = "open_preflight"
	FixRevealNode    = "reveal_node"
	FixRenameAgent   = "rename_agent"

	// Draft edits Studio performs itself.
	FixInternalChannelsOnly = "restrict_to_internal_channels"
	FixIntentGateDeny       = "set_intent_gate_deny"
	FixWriteHelperPrompt    = "write_helper_prompt"
	FixSetJoinNode          = "set_join_node"
)

// FixAction is one entry in the vocabulary.
type FixAction struct {
	ID    string `json:"id"`
	Label string `json:"label"`
	Kind  string `json:"kind"`
}

// fixActions is the whole vocabulary, in the order a UI would sensibly list it.
var fixActions = []FixAction{
	{FixOpenProviders, "Configure provider", FixKindNavigate},
	{FixOpenMCP, "Connect server", FixKindNavigate},
	{FixOpenDelivery, "Open Delivery", FixKindNavigate},
	{FixOpenSecrets, "Add the secret", FixKindNavigate},

	{FixChooseModel, "Choose a model", FixKindFocus},
	{FixAddAssertions, "Add an assertion", FixKindFocus},
	{FixRunLive, "Run it live", FixKindFocus},
	{FixOpenStudio, "Open the editor", FixKindFocus},
	{FixOpenPreflight, "Open the editor", FixKindFocus},
	{FixRevealNode, "Show the step", FixKindFocus},
	{FixRenameAgent, "Rename this agent", FixKindFocus},

	{FixInternalChannelsOnly, "Use internal channels only", FixKindApply},
	{FixIntentGateDeny, "Set the intent gate to deny", FixKindApply},
	{FixWriteHelperPrompt, "Write a starter prompt", FixKindApply},
	{FixSetJoinNode, "Set the join step", FixKindApply},
}

// FixActions returns the vocabulary. Copied so callers cannot mutate it.
func FixActions() []FixAction {
	out := make([]FixAction, len(fixActions))
	copy(out, fixActions)
	return out
}

// FixActionByID looks up one entry.
func FixActionByID(id string) (FixAction, bool) {
	id = strings.TrimSpace(id)
	for _, a := range fixActions {
		if a.ID == id {
			return a, true
		}
	}
	return FixAction{}, false
}

// FixActionLabel is the button text for an id, or "" when the id is unknown.
//
// Returning "" rather than a friendly default is deliberate: an unknown id is a
// bug — the client will not have a handler for it either — and a plausible
// label would hide that behind a button that silently does nothing.
func FixActionLabel(id string) string {
	if a, ok := FixActionByID(id); ok {
		return a.Label
	}
	return ""
}

// IsFixAction reports whether an id is in the vocabulary.
func IsFixAction(id string) bool {
	_, ok := FixActionByID(id)
	return ok
}

// FixActionIDs lists every id, for tests and for clients that want to validate.
func FixActionIDs() []string {
	out := make([]string, 0, len(fixActions))
	for _, a := range fixActions {
		out = append(out, a.ID)
	}
	return out
}

// resolveFixLabel picks the button text for a finding: its own override when it
// has one (so a finding can be specific about what its button does in context),
// otherwise the vocabulary's default.
func resolveFixLabel(action, override string) string {
	if s := strings.TrimSpace(override); s != "" {
		return s
	}
	return FixActionLabel(action)
}
